### PR TITLE
Advanced Building Blocks - Enumerables

### PR DESCRIPTION
**Pull request by cvilla714**
1. This project was based on implementing our own version of Ruby's enumerable methods
2. We created two files one that contained the methods and another one that contains the results
3.We created the following methods
- my_each
- my_each_with_index
- my_select 
- my_all
- my_any
- my_none
- my_count
- my_map
- my_inject
- multiply_els 

We implemented yield as well and created a Proc as well.

### CODE REVIEW

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Alan Soto Valle & KC Villalobos Garcia** 
Nice job with your first submission :muscle:  :muscle: 
There are some issues to solve before you can go ahead:

- [x] fix all rubocop offenses
- [x] #my_each returns an Enumerator if no block is given
    `[1,2,3].my_each.class == Enumerator` 
- [x] #my_each when a block is given when self is an array returns the array itself after calling the given block once for each element in self
        `[1,2,3,4,5].each(&proc{|n| n%2}) == [1,2,3,4,5].my_each(&proc{|n| n%2})` 
- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self
        `[1,2,3].my_each(&proc{|x| x>2}) == [1,2,3]` 
- [x] #my_each_with_index returns an enumerator if no block is given
    `[1,2,3].my_each_with_index.class == Enumerator` 
- [x] #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
        `[1,2,3,4,5].each_with_index(&proc{|n| n%2}) == [1,2,3,4,5].my_each_with_index(&proc{|n| n%2})` 
- [x] #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self
        `(1..3).my_each_with_index(&proc{|x| x>2}) == (1..3)` 
- [x] #my_select returns an array containing all elements of enum for which the given block returns a true value::range
    `(1..3).select(&proc{|x| x>2}) == (1..3).my_select(&proc{|x| x>2})` 
- [x] #my_select returns an enumerator if no block is given
    `[1,2,3,4].my_select.class==Enumerator` 
- [x] #my_all? returns true if the block never returns false or nil
    `[1,2,3].all?(&proc{|x| x>x/5}) == [1,2,3].my_all?(&proc{|x| x>x/5})` 
- [x] #my_all? returns false if the block returns false or nil
    `[1,2,3].all?(&proc{|x| x%2==0}) == [1,2,3].my_all?(&proc{|x| x%2==0}) ` 
- [x] #my_all? returns false if the block returns false or nil::range
    `(1..3).all?(&proc{|x| x%2==0}) == (1..3).my_all?(&proc{|x| x%2==0})` 
- [x] #my_all? does not mutate the original array
    `[1,2,3,4,5].tap{|t| t.all?{|n| n>3}} == [1,2,3,4,5].tap{|t| t.my_all?{|n| n>3}}` 
- [x] #my_all? when no block or argument is given returns true when none of the collection members are false or nil
      `[true, [false]].all? == [true, [false]].my_all?` 
- [x] #my_all? when no block or argument is given returns false when one of the collection members are false or nil
      `[true,[true],false].all? == [true,[true],false].my_all?` 
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
      `[1,2,3].all?(Integer) == [1,2,3].my_all?(Integer)` 
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
      `[1,-2,3.4].all?(Numeric) == [1,-2,3.4].my_all?(Numeric)` 
- [x] #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
      `['word',1,2,3].all?(Integer) == ['word',1,2,3].my_all?(Integer)` 
- [x] #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
      `['car', 'cat'].all?(/a/) == ['car', 'cat'].my_all?(/a/)` 
- [x] #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
      `['car', 'cat'].all?(/t/) == ['car', 'cat'].my_all?(/t/)` 
- [x] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
      `[5,5,5].all?(5) == [5,5,5].my_all?(5)` 
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
      `[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)` 
- [x] #my_any? returns true if the block ever returns a value other than false or nil
    `[1,2,3].any?(&proc{|x| x%2==0}) == [1,2,3].my_any?(&proc{|x| x%2==0})` 
- [x] #my_any? returns false if the block ever returns a value false or nil
    `[1,2,3].any?(&proc{|x| x==0}) == [1,2,3].my_any?(&proc{|x| x==0})` 
- [x] #my_any? returns false if the block ever returns a value false or nil::range
    `(1..3).any?(&proc{|x| x==0}) == (1..3).my_any?(&proc{|x| x==0})` 
- [x] #my_any? does not mutate the original array
    `[1,2,3,4,5].tap{|t| t.my_any?{|n| n>3}} == [1,2,3,4,5].tap{|t| t.any?{|n| n>3}}` 
- [x] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
      `[0,[]].all? == [0,[]].my_all?` 
- [x] #my_any? when no block or argument is given returns false if at least one of the collection is not true
      `[false, 0].any? == [false, 0].my_any?` 
- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
      `[1.1,'',[]].any?(Numeric) == [1.1,'',[]].my_any?(Numeric)` 
- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
      `[1,'',[]].any?(Numeric) == [1,'',[]].my_any?(Integer)` 
- [x] #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
      `['dog', 'cat'].any?(/d/) == ['dog', 'cat'].my_any?(/d/)` 
- [x] #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
      `['dog', 'cat'].any?(/z/) == ['dog', 'cat'].my_any?(/z/)` 
- [x] #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
      `['dog','car'].any?('cat') == ['dog','car'].my_any?('cat')` 
- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
      `['cat', 'dog','car'].any?('cat') == ['cat', 'dog','car'].my_any?('cat')` 
- [x] #my_none? returns true if the block never returns true for all elements
    `[1,2,3,4,5].my_none?(&proc{|n| n%7==0}) == [1,2,3,4,5].none?(&proc{|n| n%7==0})` 
- [x] #my_none? returns false if the block ever returns true for all elements
    `[1,2,3,4,5].my_none?(&proc{|n| n%2==0}) == [1,2,3,4,5].none?(&proc{|n| n%2==0})` 
- [x] #my_none? returns false if the block ever returns true for all elements::range
    `(1..3).none?(&proc{|num| num%2==0}) == (1..3).my_none?(&proc{|num| num%2==0})` 
- [x] #my_none? does not mutate the original array
    `[1,2,3,4,5].tap{|t| t.my_none?{|n| n%3}} == [1,2,3,4,5].tap{|t| t.none?{|n| n%3}}` 
- [x] #my_none? when no block or argument is given returns true only if none of the collection members is true
      `[false, nil,false].none? == [false, nil,false].my_none?` 
- [x] #my_none? when no block or argument is given returns false only if one of the collection members is true
      `[false, nil,[]].none? == [false, nil,[]].my_none?` 
- [x] #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
      `[true,[]].none?(String) == [true,[]].my_none?(String)` 
- [x] #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
      `[true,[]].none?(Numeric) == [true,[]].my_none?(Numeric)` 
- [x] #my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
      `['',[]].none?(String) == ['',[]].my_none?(String)` 
- [x] #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
      `['dog','cat'].none?(/x/) ==  ['dog','cat'].my_none?(/x/)` 
- [x] #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
      `['dog', 'cat'].none?(/d/) == ['dog', 'cat'].my_none?(/d/)` 
- [x] #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
      `['dog','car'].none?(5) == ['dog','car'].my_none?(5)` 
- [x] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
      `[5,'dog','car'].none?(5) == [5,'dog','car'].my_none?(5)` 
- [x] #my_count returns the number of items in enum through enumeration
    `[1,2,3,4,5].count == [1,2,3,4,5].my_count` 
- [x] #my_count returns the number of items in enum through enumeration::range
    `(1..3).count == (1..3).my_count` 
- [x] #my_count counts the number of items in enum that are equal to item if an argument is given
    `[1,2,3].count(1) == [1,2,3].my_count(1)` 
- [x] #my_count counts the number of elements yielding a true value if a block is given
    `[1,2,3].count(&proc{|num|num%2==0}) == [1,2,3].my_count(&proc{|num|num%2==0})` 
- [x] #my_count does not mutate the original array
    `[1,2,3,4,5].tap{|t| t.my_count{|n| n%3}} == [1,2,3,4,5].tap{|t| t.count{|n| n%3}}` 
- [x] #my_map returns a new array with the results of running a given block once for every element in enum.
    `[1,2,3].map{|x| x==2} == [1,2,3].my_map{|x| x==2}` 
- [x] #my_map returns a new array with the results of running a given block once for every element in enum.
    `[1,2,3].map(&proc{|x|x*2}) == [1,2,3].my_map(&proc{|x|x*2})` 
- [x] #my_map returns a new array with the results of running a given block once for every element in enum::range
    `(1..3).map(&proc{|num| num+1}) == (1..3).my_map(&proc{|num| num+1})` 
- [x] #my_map returns an Enumerator if no block is given
    `[1,2,3].map.class==[1,2,3].my_map.class` 
- [x] #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
       `[1,2,3].map(&proc{|x|x%2}) == [1,2,3].my_map(&proc{|x|x%2}) ` 
- [x] #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
       `[1,2,3].my_map(proc{|x|x%2}){|a|a*2} == [1,2,3].my_map(proc{|x|x%2})` 
- [x] #my_inject raises a 'LocalJumpError' when no block or argument is given
    ```
    begin
      [1,2,3,4].my_inject
    rescue LocalJumpError
      true
    end
    ```
     
- [x] #my_inject searches for the longest word in an array of strings
    `['to','four','tri'].my_inject(&proc{|a,b| [a,b].max_by{|m| m.length}}) == ['to','four','tri'].inject(&proc{|a,b| [a,b].max_by{|m| m.length}})` 
- [x] #my_inject does not mutate the original array
    ```
    array = [1,2,3,4]
    array_clone = array.clone
    array.my_inject{|total, num| total+num}
    array == array_clone
    ```
     
- [x] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
      `[1,2,3].inject(&proc{|total, num| total*num}) == [1,2,3].my_inject(&proc{|total, num| total*num})` 
- [x] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
      `(1..3).inject(&proc{|total, num| total*num}) == (1..3).my_inject(&proc{|total, num| total*num})` 
- [x] #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block
      `(1..3).inject(4) { |prod, n| prod * n } == (1..3).my_inject(4) { |prod, n| prod * n }` 
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
      `[1,2,3].inject(:+) == [1,2,3].my_inject(:+)` 
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
      `(1..9).inject(:+) == (1..9).my_inject(:+)` 
- [x] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method
      `[1,2,3].inject(4, :*) == [1,2,3].my_inject(4, :*)` 
- [x] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range
      `(1..3).inject(4, :*) == (1..3).my_inject(4, :*)` 
- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
    `multiply_els([1,2,3]) == 6`

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Hello @cvilla714 and @AlanSoto31, Great job! 👏  

There are still some changes that are necessary for this project to be approved.

Kindly make sure you effect all the changes requested by the previous TSE before always making another code review request.

- [x] Your #my_each method should handle the case when a hash is given, same goes to #my_each_with_index

- [x] You're missing #my_select method.

- [x] Your #my_all? should handle cases when a class is passed as an argument, it should return true if all of the collection is a member of such class::Integer and false if any of the collection is not. #my_any?, #my_none? should account for the same case as well as Regex.

- [x] #my_count should handle cases when a Range is given


### Kindly refer to the unresolved cases stated by the previous TSE.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: testing.rb
Comment (javitocor) [inline testing.rb]: - #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self

expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
            got: ""

File: testing.rb
Comment (javitocor) [inline testing.rb]: . #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

File: testing.rb
Comment (javitocor) [inline testing.rb]: - when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: testing.rb
Comment (javitocor) [inline testing.rb]: - #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

File: testing.rb
Comment (javitocor) [inline testing.rb]: - #my_inject raises a "LocalJumpError" when no block or argument is given

File: testing.rb
Comment (javitocor) [inline testing.rb]: Kindly remove all not used code or examples, you can create another file with the examples.

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

Hello team!
Good job implementing changes in the project! 👍 
But we still need to do some changes before getting the project approved 💪 
Kindly check my comments below 👇  implement the changes and resubmit your project for a new code review ⏩ 

## Mandatory changes

- Please, implement  #multiply_els

File: testing.rb
Comment (cvilla714) [reply testing.rb]: I don't understand your suggestion at all. The project says use multiply_els only for the inject method, not for every method, so how are we going to run the code if we comment lines 7-19

File: testing.rb
Comment (voscarmv) [inline testing.rb]: @javitocor please follow the feedback from @cvilla714  here.

When your `my_each` receives a hash and a proc in it's input, it should behave like this:

```
        enum = { a: 1, b: 2, c: 3, d: 4, e: 5 }
        my_each_output = ''
        block = proc { |num| my_each_output += num.to_s }
        enum.each(&block)
        each_output = my_each_output.dup
        my_each_output = ''
        enum.my_each(&block)

# These two variables should be the same.
        my_each_output == each_output
```

Please check that this tests passes in your computer locally before re-submitting for review.

File: testing.rb
Comment (voscarmv) [inline testing.rb]: Your `my_each_with_index` method presents the same problem as your `my_each` method. Just as before, please make sure your `my_each_with_index` is able to handle a hash with a proc in it's input.

Here's the error message returned by the automated tests:

```
 2) enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)
     
       expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
            got: "Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\n"
     
       (compared using ==)
     
       Diff:
       @@ -1,6 +1,6 @@
       -Num: [:a, 1], idx: 0
       -Num: [:b, 2], idx: 1
       -Num: [:c, 3], idx: 2
       -Num: [:d, 4], idx: 3
       -Num: [:e, 5], idx: 4
       +Num: , idx: 0
       +Num: , idx: 1
       +Num: , idx: 2
       +Num: , idx: 3
       +Num: , idx: 4
```

File: testing.rb
Comment (voscarmv) [inline testing.rb]: Please make sure your `my_none` passes the following test:

```
words = %w[dog door rod blade]

# This should return true
words.my_none?(5) == true

words2 = %w[5 dog door rod blade]
# This should return false
words2.my_none?(5) == true

```

File: testing.rb
Comment (voscarmv) [inline testing.rb]: Please make sure your `my_inject` returns a proper error message. Here's the error from our automated tests:

```
5) enumerables #my_inject raises a "LocalJumpError" when no block or argument is given
     Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)
       expected LocalJumpError but nothing was raised
```

Find out how to raise a `LocalJumpError` in ruby and implement it as requested in the error message.

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Nice work so far, team. 👍 
## However, there are some changes required 👀 🧰 🔨 🔧 

- Nex time if you don't understand the feedback from the previous TSE, please request for help on Slack before re-submitting for review without folloring the TSE's instructions!
- Please read my code comments for feedback on fixing your code 👍 

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

File: testing.rb
Comment (acushlakoncept) [inline testing.rb]: Your #my_each method should return the range itself after calling the given block when a block is given and self is a range.  
If you notice, line 10 - 12 converts both hash or range to an array,  but if a range/hash is given as a block it should return a range/hash as well.... changing line 18 to `self` should fix this.
This applies to #my_each_with_index as well.

File: testing.rb
Comment (acushlakoncept) [inline testing.rb]: Your #my_map does not handle cases when a `proc` is given:
Your method should look like this: `def my_map(proc = nil)`
Within the method add the logic that handles a proc if given. [See docs](https://ruby-doc.org/core-2.6/Proc.html) about proc

File: testing.rb
Comment (acushlakoncept) [inline testing.rb]: Your #multiply_els accepts an **array as an argument** and multiplies all the elements of the array together using **#my_inject**  So you should have something like `def multiply_els(arr)`

You don't need all those codes inside your #multiply_els method, call the #my_inject method inside.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Hello @cvilla714 and @AlanSoto31, Great job! 👏  Both of you have done well on this project.
Let's make a few changes, and we will be done.

Kindly check my comments below on your codebase, fix those and you should be done.

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You guys have done an excellent job so far. We have just two(2) more changes to make and we should be good to go. Let's check it out:__

**[ REQUIRED CHANGES ]**

- [x] Your GitHub Actions linter checks are not running. 
- It looks like you have it set up, but for a reason, it's not running on your pull request page. Kindly fix this.
- Make sure you fix all linter errors before you submit another review request.

- [x] You need to refactor your `multiply_els` method.
- You should move your `multiply_els` method out of the Enumerable module. This is because this method is not part of the enumerable methods. It's just there to test your `my_inject` method. Therefore, let it stand alone.



- Secondly, you should not have so much logic in your `multiply_els` method. Since the main goal of this method is to test the `my_inject` method, all you have to do is call the `my_inject` method inside this method and pass a block that works the array. You can refactor your `multiply_els` method to be as simple as the snippet below:

```
def multiply_els(arr)
  arr.my_inject { |result, item| result * item }
end
```
I believe this should work. You can test it with the data given on the project description page, just to be sure.


**[ OPTIONAL CHANGES ]**

- As mentioned by one of the previous reviewers, the `results.rb` file is a test file you used in development. This file should not necessarily be committed. 
- If you wish to keep this file locally, you can remove this file from your project remote repository by:
  - Copy the file content to somewhere safe
  - Delete the file locally
  - Commit your changes
  - Create a `.gitignore` file and add the name (`results.rb`)
  - Add a new file with the file name (`results.rb`) and paste the content.
  - Commit your changes.

With this, you should have your file available and accessible locally, and it will not be committed to your remote repository.

**[ COMMENTS ]**
- If you carefully follow my review messages, your project should be approved upon your next review request.
- Feel free to reach out if you're having any questions.

* * * 
**Please make the required changes and submit another review request when you are done.**

* * *

**_You can reach out to me on Twitter (@ola_lere), Github(@teekaytech), or slack (@taofeekolalere )_**
__Happy Coding :smiley:__

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Hello @cvilla714 and @AlanSoto31,

You've done an awesome job so far 👌.
Let us quickly fix one last method and request another review.

- [x] Your `#multiply_els` is still failing, fix this and you should be good to go: `multiply_els([2, 2, 3, 2])` should return 24 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by maelfosso — APPROVED** [review]
# Congratulations :clap: 

Thank you for implementing all the suggested changes :blush: 
It's OK correctly. You can merge it now :+1: 

Good continuation :handshake: 

Mael FOSSO

**Pull request merged**
